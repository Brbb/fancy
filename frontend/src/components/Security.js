import React, { Component } from "react";
import "./Settings.css";

class Security extends Component {
  
  render() {
    return (
      <div>
      
      </div>
    );
  }
}

export default Security;
