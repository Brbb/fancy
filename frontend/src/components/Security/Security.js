import React, { Component } from "react";
import { Button, InputField, Message } from "../Elements/Basics";
import PasswordGroup from "../Elements/PasswordGroup";
import api from "../../services/users/api";
import { withRouter } from "react-router-dom";

class Security extends Component {
  constructor(props) {
    super(props);
    this.state = {
      oldPassword: "",
      newPassword: "",
      repeatNewPassword: "",
      onUserSettingsChange: () => {},
      error: ""
    };
  }

  handleOldPasswordChange = event => {
    this.setState({ oldPassword: event.target.value });
  };

  save = async () => {
    let result = await api.updatePassword(
      this.props.user,
      this.state.oldPassword,
      this.state.newPassword,
      this.state.repeatNewPassword
    );

    // logout
    if (!result.err) {
      await this.props.onUserSettingsChange();
    } else {
      this.setState({
        error: result.err
      });
    }
  };

  delete = async () => {
    let result = await api.delete(this.props.user);
    if (!result.err) {
      await this.props.onUserSettingsChange();
    } else {
      this.setState({
        error: "User not deleted!"
      });
    }
  };

  render() {
    return (
      <div>
        <div className="f-section">
          <label className="f-section-title-no-margin">Change Password</label>
          <hr className="f-hr-section" />
          <InputField
            name="oldPasswordField"
            type="password"
            placeholder="Old password"
            value={this.state.oldPassword}
            onChange={this.handleOldPasswordChange}
            text="Old Password"
          />
          <PasswordGroup
            onPasswordsChange={(newPassword, repeatPassword, error) => {
              this.setState({
                newPassword: newPassword,
                repeatNewPassword: repeatPassword,
                error: error
              });
            }}
          />
          <Button
            name="save-button"
            className="f-submit-button"
            text="Save Changes"
            onClick={this.save}
            disabled={
              this.state.oldPassword === "" ||
              this.state.newPassword === "" ||
              this.state.repeatNewPassword === "" ||
              this.state.newPassword !== this.state.repeatNewPassword
            }
          />
          <Message text={this.state.error} success={this.state.error === ""} />
        </div>
        <div className="f-section">
          <label className="f-section-title-no-margin warning-text">
            Delete Account
          </label>
          <hr className="f-hr-section" />
          <p className="f-description">
            Once the deletion process begins you will not be able to access to
            your profile and saved settings.
          </p>
          <Button
            name="delete-button"
            className="f-button warning"
            text="Delete"
            onClick={() => {
              if (
                window.confirm(
                  "Are you sure you want to delete your profile?\n(You will be redirected to the login screen)"
                )
              )
                return this.delete();
            }}
          />
        </div>
      </div>
    );
  }
}

export default withRouter(Security);
