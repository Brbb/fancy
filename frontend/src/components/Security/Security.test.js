import Security from "./Security";
import userApi from "../../services/users/api";

describe("Security component", () => {
  let mockUser = {
    settings: {
      language: "Chinese",
      isPrivate: false
    },
    email: "asd",
    _id: "obj1234"
  };

  describe("Rendering tests", () => {
    test("Security mounts and fields are initialized correctly", () => {
      let wrapper = shallow(<Security.WrappedComponent user={mockUser} />);
      let saveButton = wrapper.find('[name="save-button"]').first();

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.state().oldPassword).toEqual("");
      expect(wrapper.state().newPassword).toEqual("");
      expect(wrapper.state().repeatNewPassword).toEqual("");
      expect(saveButton.prop("disabled")).toBeTruthy();
    });

    test("Save button gets enabled correctly on field/state changes", () => {
      let wrapper = mount(<Security.WrappedComponent user={mockUser} />);
      wrapper
        .find('[name="oldPasswordField"]')
        .last()
        .simulate("change", { target: { value: "old" } });
      wrapper
        .find('[name="newPasswordField"]')
        .last()
        .simulate("change", { target: { value: "new" } });
      wrapper
        .find('[name="repeatNewPasswordField"]')
        .last()
        .simulate("change", { target: { value: "new" } });

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.state().oldPassword).toEqual("old"); 
      expect(wrapper.state().newPassword).toEqual("new");
      expect(wrapper.state().repeatNewPassword).toEqual("new");

      // re-find the button to include re-rendered prop
      let saveButton = wrapper.find('[name="save-button"]').first();
      expect(saveButton.prop("disabled")).toBeFalsy();

      wrapper
        .find('[name="repeatNewPasswordField"]')
        .last()
        .simulate("change", { target: { value: "wrong" } });
      expect(wrapper.state().repeatNewPassword).toEqual("wrong");
      expect(wrapper.state().error).toEqual("Passwords must match");
    });
  });

  describe("Behavioral test", () => {
    test("Save button calls updatePassword and on success logs out", () => {
      let logout = jest.fn();

      let wrapper = shallow(
        <Security.WrappedComponent
          user={mockUser}
          onUserSettingsChange={logout}
        />
      );
      jest.spyOn(userApi, "updatePassword").mockImplementation(() => {
        return Promise.resolve({});
      });

      expect(logout).toBeCalledTimes(0);
      expect(userApi.updatePassword).toBeCalledTimes(0);

      let saveButton = wrapper.find('[name="save-button"]').first();
      return saveButton
        .prop("onClick")()
        .then(() => {
          expect(userApi.updatePassword).toBeCalledTimes(1);
          expect(logout).toBeCalledTimes(1);
          jest.clearAllMocks();
        });
    });

    test("Save button calls updatePassword and on error shows a message", () => {
      let logout = jest.fn();

      let wrapper = shallow(
        <Security.WrappedComponent
          user={mockUser}
          onUserSettingsChange={logout}
        />
      );
      jest.spyOn(userApi, "updatePassword").mockImplementation(() => {
        return Promise.resolve({ err: "any" });
      });

      expect(wrapper.state().error).toBe("");
      expect(logout).toBeCalledTimes(0);
      expect(userApi.updatePassword).toBeCalledTimes(0);

      let saveButton = wrapper.find('[name="save-button"]').first();
      return saveButton
        .prop("onClick")()
        .then(() => {
          expect(userApi.updatePassword).toBeCalledTimes(1);
          expect(logout).toBeCalledTimes(0);
          expect(wrapper.state().error).toBe("any");
          jest.clearAllMocks();
        });
    });

    test("Delete button calls delete and on success logs out", () => {
      let logout = jest.fn();

      let wrapper = shallow(
        <Security.WrappedComponent
          user={mockUser}
          onUserSettingsChange={logout}
        />
      );
      jest.spyOn(userApi, "delete").mockImplementation(() => {
        return Promise.resolve({});
      });
      jest.spyOn(window, "confirm").mockReturnValue(true);

      expect(logout).toBeCalledTimes(0);
      expect(userApi.delete).toBeCalledTimes(0);

      let deleteButton = wrapper.find('[name="delete-button"]').first();
      return deleteButton
        .prop("onClick")()
        .then(() => {
          expect(userApi.delete).toBeCalledTimes(1);
          expect(logout).toBeCalledTimes(1);
          jest.clearAllMocks();
        });
    });

    test("Delete button calls delete and on error shows a message", () => {
      let logout = jest.fn();

      let wrapper = shallow(
        <Security.WrappedComponent
          user={mockUser}
          onUserSettingsChange={logout}
        />
      );

      jest.spyOn(window, "confirm").mockReturnValue(true);
      jest.spyOn(userApi, "delete").mockImplementation(() => {
        return Promise.resolve({ err: "any" });
      });

      expect(wrapper.state().error).toBe("");
      expect(logout).toBeCalledTimes(0);
      expect(userApi.delete).toBeCalledTimes(0);

      let deleteButton = wrapper.find('[name="delete-button"]').first();
      return deleteButton
        .prop("onClick")()
        .then(() => {
          expect(userApi.delete).toBeCalledTimes(1);
          expect(logout).toBeCalledTimes(0);
          expect(wrapper.state().error).toBe("User not deleted!");
          jest.clearAllMocks();
        });
    });
  });
});
