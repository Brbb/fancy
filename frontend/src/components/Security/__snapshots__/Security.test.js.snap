// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Security component Rendering tests Save button gets enabled correctly on field/state changes 1`] = `
<div>
  <div
    className="f-section"
  >
    <label
      className="f-section-title-no-margin"
    >
      Change Password
    </label>
    <hr
      className="f-hr-section"
    />
    <div
      className="f-input"
    >
      <label
        className="f-input-field-label"
      >
        Old Password
      </label>
      <input
        className="f-input-field"
        name="oldPasswordField"
        onChange={[Function]}
        placeholder="Old password"
        type="password"
        value="old"
      />
    </div>
    <div>
      <div
        className="f-input"
      >
        <label
          className="f-input-field-label"
        >
          Password
        </label>
        <input
          className="f-input-field"
          name="newPasswordField"
          onChange={[Function]}
          placeholder="Super Secret"
          type="password"
          value="new"
        />
      </div>
      <div
        className="f-input"
      >
        <label
          className="f-input-field-label"
        />
        <input
          className="f-input-field"
          name="repeatNewPasswordField"
          onChange={[Function]}
          placeholder="Repeat Super Secret"
          type="password"
          value="new"
        />
      </div>
    </div>
    <input
      className="f-submit-button"
      disabled={false}
      name="save-button"
      onClick={[Function]}
      type="button"
      value="Save Changes"
    />
    <label
      className="f-message f-message-s"
    />
  </div>
  <div
    className="f-section"
  >
    <label
      className="f-section-title-no-margin warning-text"
    >
      Delete Account
    </label>
    <hr
      className="f-hr-section"
    />
    <p
      className="f-description"
    >
      Once the deletion process begins you will not be able to access to your profile and saved settings.
    </p>
    <input
      className="f-button warning"
      name="delete-button"
      onClick={[Function]}
      type="button"
      value="Delete"
    />
  </div>
</div>
`;

exports[`Security component Rendering tests Security mounts and fields are initialized correctly 1`] = `
<div>
  <div
    className="f-section"
  >
    <label
      className="f-section-title-no-margin"
    >
      Change Password
    </label>
    <hr
      className="f-hr-section"
    />
    <InputField
      name="oldPasswordField"
      onChange={[Function]}
      placeholder="Old password"
      text="Old Password"
      type="password"
      value=""
    />
    <PasswordGroup
      onPasswordsChange={[Function]}
    />
    <Button
      className="f-submit-button"
      disabled={true}
      name="save-button"
      onClick={[Function]}
      text="Save Changes"
    />
    <Message
      success={true}
      text=""
    />
  </div>
  <div
    className="f-section"
  >
    <label
      className="f-section-title-no-margin warning-text"
    >
      Delete Account
    </label>
    <hr
      className="f-hr-section"
    />
    <p
      className="f-description"
    >
      Once the deletion process begins you will not be able to access to your profile and saved settings.
    </p>
    <Button
      className="f-button warning"
      name="delete-button"
      onClick={[Function]}
      text="Delete"
    />
  </div>
</div>
`;
